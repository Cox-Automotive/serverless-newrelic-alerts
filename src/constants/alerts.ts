export enum FunctionAlert {
  DURATION_1_SEC = 'functionDuration1Sec',
  ERRORS = 'functionErrors',
  THROTTLES = 'functionThrottles'
}

export enum ApiGatewayAlert {
  ERRORS_4XX = 'apiGateway4XXErrors',
  ERRORS_5XX = 'apiGateway5XXErrors'
}

export enum SqsAlert {
  DLQ_VISIBLE_MESSAGES = 'sqsDlqVisibleMessages'
}

export enum DynamoDbAlert {
  BATCH_GET_SYSTEM_ERRORS = 'dynamoDbBatchGetSystemErrors',
  BATCH_WRITE_SYSTEM_ERRORS = 'dynamoDbBatchWriteSystemErrors',
  DELETE_SYSTEM_ERRORS = 'dynamoDbDeleteSystemErrors',
  GET_SYSTEM_ERRORS = 'dynamoDbGetSystemErrors',
  PUT_SYSTEM_ERRORS = 'dynamoDbPutSystemErrors',
  QUERY_SYSTEM_ERRORS = 'dynamoDbQuerySystemErrors',
  SCAN_SYSTEM_ERRORS = 'dynamoDbScanSystemErrors',
  UPDATE_SYSTEM_ERRORS = 'dynamoDbUpdateSystemErrors',
  USER_ERRORS = 'dynamoDbUserErrors'
}
