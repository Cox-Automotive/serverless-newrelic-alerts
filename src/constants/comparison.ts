export enum Comparison {
  ABOVE = 'above',
  BELOW = 'below',
  EQUAL = 'equal'
}
