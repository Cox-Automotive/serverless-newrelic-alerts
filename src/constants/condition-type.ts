export enum ConditionType {
  INFRA_METRIC = 'infra_metric'
}
