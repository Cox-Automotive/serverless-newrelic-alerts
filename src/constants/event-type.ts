export enum EventType {
  SERVERLESS_SAMPLE = 'ServerlessSample',
  API_GATEWAY_SAMPLE = 'ApiGatewaySample',
  QUEUE_SAMPLE = 'QueueSample'
}
