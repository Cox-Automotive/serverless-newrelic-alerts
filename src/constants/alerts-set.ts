import { DynamoDbAlert } from './alerts'

export enum AlertsSet {
  DYNAMO_DB_SYSTEM_ERRORS = 'dynamoDbSystemErrors'
}

export const AlertsSetMapping = {
  [AlertsSet.DYNAMO_DB_SYSTEM_ERRORS]: [
    DynamoDbAlert.BATCH_GET_SYSTEM_ERRORS,
    DynamoDbAlert.BATCH_WRITE_SYSTEM_ERRORS,
    DynamoDbAlert.DELETE_SYSTEM_ERRORS,
    DynamoDbAlert.GET_SYSTEM_ERRORS,
    DynamoDbAlert.PUT_SYSTEM_ERRORS,
    DynamoDbAlert.QUERY_SYSTEM_ERRORS,
    DynamoDbAlert.SCAN_SYSTEM_ERRORS,
    DynamoDbAlert.UPDATE_SYSTEM_ERRORS
  ]
}
