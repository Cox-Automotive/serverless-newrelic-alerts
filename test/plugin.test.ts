import Serverless from 'serverless'
import { ApiGatewayAlert, DynamoDbAlert, FunctionAlert, SqsAlert } from '../src/constants/alerts'
import { AlertsSet } from '../src/constants/alerts-set'

const NewRelicAlertsPlugin = require('../src')

describe('Newrelic Alert Plugin', () => {
  const logMock = jest.fn()
  const getServerless = (config: any, { resources, functions = [] }: any = {}): Serverless =>
    ({
      service: {
        custom: {
          newrelicAlerts: config
        },
        provider: {
          compiledCloudFormationTemplate: {
            Resources: resources
          }
        },
        getServiceName() {
          return 'test service'
        },
        getAllFunctionsNames() {
          return functions.map(({ displayName }) => displayName)
        },
        getAllFunctions() {
          return functions.map(({ name }) => name)
        },
        getFunction(functionName) {
          const fn = functions.find(({ name }) => name === functionName)
          return {
            alerts: fn.alerts,
            name: fn.displayName
          }
        }
      },
      getProvider() {
        return {
          getStage() {
            return 'test'
          }
        }
      },
      cli: {
        log: logMock
      }
    } as any)

  const minimalConfig = {
    policyServiceToken: 'policy-token',
    infrastructureConditionServiceToken: 'infrastructure-condition-token'
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('constructor', () => {
    it('should throw error if some of tokens are not provided', () => {
      expect(() => new NewRelicAlertsPlugin(getServerless({}))).toThrow()
    })

    it("shouldn't apply hooks if there is no plugin config", () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(undefined))
      expect(plugin.hooks).toEqual({})
    })
  })

  describe('getPolicyCloudFormation', () => {
    it('should generate valid policy', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      const policy = plugin.getPolicyCloudFormation()
      expect(policy).toMatchSnapshot()
    })
  })

  describe('getInfrastructureConditionCloudFormation', () => {
    it('should generate valid infrastructure condition', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      const infrastructureCondition = plugin.getInfrastructureConditionCloudFormation(
        FunctionAlert.THROTTLES,
        ['fn-1', 'fn-2']
      )
      expect(infrastructureCondition).toMatchSnapshot()
    })
  })

  describe('checkEligibility', () => {
    it('should return true if there are both resources and alerts', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      expect(plugin.checkEligibility([FunctionAlert.THROTTLES], ['some res'])).toBe(true)
    })

    it('should return false if there are no alerts', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      expect(plugin.checkEligibility([], ['some res'])).toBe(false)
    })

    it('should return false if there are no resources', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      expect(plugin.checkEligibility([], [])).toBe(false)
    })

    it('should log warn if there are no resources for alerts configured', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      plugin.checkEligibility([FunctionAlert.THROTTLES], [])
      expect(logMock).toBeCalled()
    })
  })

  describe('getFunctionAlertsCloudFormation', () => {
    it('should generate global alerts for all functions', () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: Object.values(FunctionAlert)
          },
          {
            functions: [
              {
                name: 'test-function',
                displayName: 'test-service-test-function'
              },
              {
                name: 'test-function-2',
                displayName: 'test-service-test-function-2'
              }
            ]
          }
        )
      )
      const cf = plugin.getFunctionAlertsCloudFormation()
      expect(cf).toMatchSnapshot()
    })

    it("shouldn't generate global alerts for function if alert disabled locally", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: Object.values(FunctionAlert)
          },
          {
            functions: [
              {
                name: 'test-function',
                displayName: 'test-service-test-function'
              },
              {
                name: 'test-function-2',
                displayName: 'test-service-test-function-2',
                alerts: [
                  {
                    name: FunctionAlert.THROTTLES,
                    enabled: false
                  }
                ]
              }
            ]
          }
        )
      )
      const cf = plugin.getFunctionAlertsCloudFormation()
      expect(cf.FunctionThrottlesInfrastructureCondition.Properties.data.filter).toMatchSnapshot()
    })

    it('should generate local defined alerts functions', () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig
          },
          {
            functions: [
              {
                name: 'test-function',
                displayName: 'test-service-test-function'
              },
              {
                name: 'test-function-2',
                displayName: 'test-service-test-function-2',
                alerts: [FunctionAlert.DURATION_1_SEC]
              }
            ]
          }
        )
      )
      const cf = plugin.getFunctionAlertsCloudFormation()
      expect(cf.FunctionDuration1SecInfrastructureCondition).toBeDefined()
      expect(
        cf.FunctionDuration1SecInfrastructureCondition.Properties.data.filter
      ).toMatchSnapshot()
    })

    it("shouldn't fail if there is no functions", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless({
          ...minimalConfig,
          alerts: [...Object.values(FunctionAlert), ...Object.values(ApiGatewayAlert)]
        })
      )
      const cf = plugin.getFunctionAlertsCloudFormation()
      expect(cf).toEqual({})
    })

    it("shouldn't fail if there is no alerts", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: Object.values(ApiGatewayAlert)
          },
          {
            functions: [
              {
                name: 'test-function',
                displayName: 'test-service-test-function'
              }
            ]
          }
        )
      )
      const cf = plugin.getFunctionAlertsCloudFormation()
      expect(cf).toEqual({})
    })
  })

  describe('getApiGatewayAlertsCloudFormation', () => {
    it('should generate alerts for all api gateways', () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: [...Object.values(FunctionAlert), ...Object.values(ApiGatewayAlert)]
          },
          {
            resources: {
              ApiGateway: {
                Type: 'AWS::ApiGateway::RestApi',
                Properties: {
                  Name: 'api-gatway'
                }
              },
              ApiGateway2: {
                Type: 'AWS::ApiGateway::RestApi',
                Properties: {
                  Name: 'api-gatway2'
                }
              }
            }
          }
        )
      )
      const cf = plugin.getApiGatewayAlertsCloudFormation()
      expect(cf).toMatchSnapshot()
    })

    it("shouldn't fail if there is no api gateways", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless({
          ...minimalConfig,
          alerts: [...Object.values(FunctionAlert), ...Object.values(ApiGatewayAlert)]
        })
      )
      const cf = plugin.getApiGatewayAlertsCloudFormation()
      expect(cf).toEqual({})
    })

    it("shouldn't fail if there is no alerts", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: Object.values(FunctionAlert)
          },
          {
            resources: {
              ApiGateway: {
                Type: 'AWS::ApiGateway::RestApi',
                Properties: {
                  Name: 'api-gatway'
                }
              }
            }
          }
        )
      )
      const cf = plugin.getApiGatewayAlertsCloudFormation()
      expect(cf).toEqual({})
    })
  })

  describe('getSqsAlertsCloudFormation', () => {
    it('should generate alerts for all dead letter queues', () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: [...Object.values(DynamoDbAlert), SqsAlert.DLQ_VISIBLE_MESSAGES]
          },
          {
            resources: {
              Queue: {
                Type: 'AWS::SQS::Queue',
                Properties: {
                  QueueName: 'simple-queue'
                }
              },
              QueueDlq: {
                Type: 'AWS::SQS::Queue',
                Properties: {
                  QueueName: 'simple-queue-dlq'
                }
              }
            }
          }
        )
      )
      const cf = plugin.getSqsAlertsCloudFormation()
      expect(cf).toMatchSnapshot()
    })

    it("shouldn't fail if there is no dead letter queues", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: [SqsAlert.DLQ_VISIBLE_MESSAGES, ...Object.values(ApiGatewayAlert)]
          },
          {
            resources: {
              Queue: {
                Type: 'AWS::SQS::Queue',
                Properties: {
                  QueueName: 'simple-queue'
                }
              }
            }
          }
        )
      )
      const cf = plugin.getSqsAlertsCloudFormation()
      expect(cf).toEqual({})
    })

    it("shouldn't fail if there is no alerts", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: Object.values(FunctionAlert)
          },
          {
            resources: {
              QueueDlq: {
                Type: 'AWS::SQS::Queue',
                Properties: {
                  QueueName: 'simple-queue-dlq'
                }
              }
            }
          }
        )
      )
      const cf = plugin.getSqsAlertsCloudFormation()
      expect(cf).toEqual({})
    })
  })

  describe('getDynamoDbAlertsCloudFormation', () => {
    it('should generate alerts for all dynamo tables', () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: [...Object.values(DynamoDbAlert), ...Object.values(ApiGatewayAlert)]
          },
          {
            resources: {
              DynamoDBTable: {
                Type: 'AWS::DynamoDB::Table',
                Properties: {
                  TableName: 'dynamo-table'
                }
              },
              DynamoDBTable2: {
                Type: 'AWS::DynamoDB::Table',
                Properties: {
                  TableName: 'dynamo-table2'
                }
              }
            }
          }
        )
      )
      const cf = plugin.getDynamoDbAlertsCloudFormation()
      expect(cf).toMatchSnapshot()
    })

    it("shouldn't fail if there is no dynamo tables", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless({
          ...minimalConfig,
          alerts: [...Object.values(DynamoDbAlert), ...Object.values(ApiGatewayAlert)]
        })
      )
      const cf = plugin.getDynamoDbAlertsCloudFormation()
      expect(cf).toEqual({})
    })

    it("shouldn't fail if there is no alerts", () => {
      const plugin = new NewRelicAlertsPlugin(
        getServerless(
          {
            ...minimalConfig,
            alerts: Object.values(FunctionAlert)
          },
          {
            resources: {
              DynamoDB: {
                Type: 'AWS::DynamoDB::Table',
                Properties: {
                  Name: 'dynamo-table'
                }
              }
            }
          }
        )
      )
      const cf = plugin.getDynamoDbAlertsCloudFormation()
      expect(cf).toEqual({})
    })
  })

  describe('getAlerts', () => {
    it('should spread all alerts set', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      const alerts = plugin.getAlerts([AlertsSet.DYNAMO_DB_SYSTEM_ERRORS])
      expect(alerts).toEqual([
        DynamoDbAlert.BATCH_GET_SYSTEM_ERRORS,
        DynamoDbAlert.BATCH_WRITE_SYSTEM_ERRORS,
        DynamoDbAlert.DELETE_SYSTEM_ERRORS,
        DynamoDbAlert.GET_SYSTEM_ERRORS,
        DynamoDbAlert.PUT_SYSTEM_ERRORS,
        DynamoDbAlert.QUERY_SYSTEM_ERRORS,
        DynamoDbAlert.SCAN_SYSTEM_ERRORS,
        DynamoDbAlert.UPDATE_SYSTEM_ERRORS
      ])
      expect(alerts).not.toContain(AlertsSet.DYNAMO_DB_SYSTEM_ERRORS)
    })

    it('should filter out with warning all unknown alerts', () => {
      const plugin = new NewRelicAlertsPlugin(getServerless(minimalConfig))
      const alerts = plugin.getAlerts([FunctionAlert.THROTTLES, 'unknownAlert'])
      expect(alerts).toContain(FunctionAlert.THROTTLES)
      expect(alerts).not.toContain('unknownAlert')
    })
  })
})
