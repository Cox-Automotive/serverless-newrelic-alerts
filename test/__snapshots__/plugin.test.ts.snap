// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newrelic Alert Plugin getApiGatewayAlertsCloudFormation should generate alerts for all api gateways 1`] = `
Object {
  "ApiGateway4XxErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "ApiGatewaySample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "apiName": Array [
                  "api-gatway",
                  "api-gatway2",
                ],
              },
            },
          ],
        },
        "integration_provider": undefined,
        "name": "Api Gateway 4 XX Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.4xxError.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "ApiGateway5XxErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "ApiGatewaySample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "apiName": Array [
                  "api-gatway",
                  "api-gatway2",
                ],
              },
            },
          ],
        },
        "integration_provider": undefined,
        "name": "Api Gateway 5 XX Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.5xxError.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
}
`;

exports[`Newrelic Alert Plugin getDynamoDbAlertsCloudFormation should generate alerts for all dynamo tables 1`] = `
Object {
  "DynamoDbBatchGetSystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Batch Get System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.batchGetSystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbBatchWriteSystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Batch Write System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.batchWriteSystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbDeleteSystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Delete System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.deleteSystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbGetSystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Get System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.getSystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbPutSystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Put System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.putSystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbQuerySystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Query System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.querySystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbScanSystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Scan System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.scanSystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbUpdateSystemErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db Update System Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.updateSystemErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "DynamoDbUserErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "DatastoreSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "tableName": Array [
                  "dynamo-table",
                  "dynamo-table2",
                ],
              },
            },
          ],
        },
        "integration_provider": "DynamoDbTable",
        "name": "Dynamo Db User Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.userErrors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
}
`;

exports[`Newrelic Alert Plugin getFunctionAlertsCloudFormation should generate global alerts for all functions 1`] = `
Object {
  "FunctionDuration1SecInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 1000,
        },
        "enabled": true,
        "event_type": "ServerlessSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "displayName": Array [
                  "test-service-test-function",
                  "test-service-test-function-2",
                ],
              },
            },
          ],
        },
        "integration_provider": "LambdaFunction",
        "name": "Function Duration 1 Sec",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.duration.Maximum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "FunctionErrorsInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "ServerlessSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "displayName": Array [
                  "test-service-test-function",
                  "test-service-test-function-2",
                ],
              },
            },
          ],
        },
        "integration_provider": "LambdaFunction",
        "name": "Function Errors",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.errors.Sum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
  "FunctionThrottlesInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "ServerlessSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "displayName": Array [
                  "test-service-test-function",
                  "test-service-test-function-2",
                ],
              },
            },
          ],
        },
        "integration_provider": "LambdaFunction",
        "name": "Function Throttles",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.throttles.Maximum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
}
`;

exports[`Newrelic Alert Plugin getFunctionAlertsCloudFormation should generate local defined alerts functions 1`] = `
Object {
  "TestFunction2FunctionDuration1SecInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 1000,
        },
        "enabled": true,
        "event_type": "ServerlessSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "displayName": Array [
                  "test-service-test-function-2",
                ],
              },
            },
          ],
        },
        "integration_provider": "LambdaFunction",
        "name": "Test Function 2 Function Duration 1 Sec",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.duration.Maximum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
}
`;

exports[`Newrelic Alert Plugin getPolicyCloudFormation should generate valid policy 1`] = `
Object {
  "TestServiceTestNewRelicPolicy": Object {
    "Properties": Object {
      "ServiceToken": "policy-token",
      "policy": Object {
        "incident_preference": undefined,
        "name": "TestService TEST",
      },
    },
    "Type": "Custom::NewRelicPolicy",
  },
}
`;

exports[`Newrelic Alert Plugin getPolicyCloudFormation should generate valid policy with custom name 1`] = `
Object {
  "CustomPolicyNameTestNewRelicPolicy": Object {
    "Properties": Object {
      "ServiceToken": "policy-token",
      "policy": Object {
        "incident_preference": undefined,
        "name": "Custom policy name TEST",
      },
    },
    "Type": "Custom::NewRelicPolicy",
  },
}
`;

exports[`Newrelic Alert Plugin getSqsAlertsCloudFormation should generate alerts for all dead letter queues 1`] = `
Object {
  "SqsDlqVisibleMessagesInfrastructureCondition": Object {
    "Properties": Object {
      "ServiceToken": "infrastructure-condition-token",
      "data": Object {
        "comparison": "above",
        "critical_threshold": Object {
          "duration_minutes": 5,
          "time_function": "all",
          "value": 3,
        },
        "enabled": true,
        "event_type": "QueueSample",
        "filter": Object {
          "and": Array [
            Object {
              "in": Object {
                "queueName": Array [
                  "simple-queue-dlq",
                ],
              },
            },
          ],
        },
        "integration_provider": "SqsQueue",
        "name": "Sqs Dlq Visible Messages",
        "policy_id": Object {
          "Ref": "TestServiceTestNewRelicPolicy",
        },
        "runbook_url": undefined,
        "select_value": "provider.approximateNumberOfMessagesVisible.Maximum",
        "type": "infra_metric",
        "violation_close_timer": 24,
      },
      "policy_id": Object {
        "Ref": "TestServiceTestNewRelicPolicy",
      },
    },
    "Type": "Custom::NewRelicInfrastructureCondition",
  },
}
`;
